File: Stacks&Queues/StacksAndQueues.py


"""
========================================================================
Stacks & Queues

STACK: LIFO:    Implement a stack using a linked list.
                Using position 0 in a linked list we achieve:
                    add:    O(1)
                    remove: O(1)

QUEUE: FIFO:    Implement a queue using a linked list.
                    add:    O(1)
                    remove: O(1)
      
: zach wolpe
: zach.wolpe@medibio.com.au
: 30 June 2023
========================================================================
"""

class Node:
    def __init__(self, value) -> None:
        self.value = value
        self.next  = None

class Stack:
    def __init__(self, value) -> None:
        node = Node(value)
        self.top    = node
        self.height = 1

    def __str__(self) -> str:
        msg = '\nStack: ['
        node = self.top
        while node:
            msg += str(node.value) + ','
            node = node.next
        if self.height > 0:
            msg = msg[:-1]
        msg += ']'
        return msg
    
    def push(self, value) -> None:
        node = Node(value)
        if self.height == 0:
            self.top = node
        else:
            node.next = self.top
            self.top = node
        self.height += 1
        
    def pop(self) -> Node:
        if self.height == 0:
            return None
        node = self.top
        self.top = self.top.next
        node.next = None
        self.height -= 1
        return node


class Queue:
    def __init__(self, value) -> None:
        node = Node(value)
        self.first  = node
        self.last   = node
        self.length = 1
    
    def __str__(self) -> str:
        msg = '\nQueue: ['
        node = self.first
        while node:
            msg += str(node.value) + ','
            node = node.next
        if self.length > 0:
            msg = msg[:-1]
        msg += ']'
        return msg

    def enqueue(self, value) -> None:
        node = Node(value)
        if self.length == 0:
            self.first = node
            self.last  = node
        else:
            self.last.next = node
            self.last = node
        self.length += 1

    def dequeue(self) -> Node:
        if self.length == 0:
            return None
        node = self.first
        if self.length == 1:
            self.first = None
            self.last  = None
        else:
            self.first = self.first.next
            node.next = None
        self.length -= 1
        return node


if __name__ == '__main__':
    queue = Queue(1)
    queue.enqueue(22)
    print(queue)
    print('dequeue: ', queue.dequeue().value)
    print(queue)
    queue.dequeue()
    print(queue)
    queue.dequeue()
    print(queue)


    # stack = Stack(2)
    # stack.push(11)
    # stack.push(22)
    # stack.push(33)
    # print(stack)
    # stack.pop()
    # print(stack)
    # stack.pop()
    # stack.pop()
    # stack.pop()
    # print(stack)

File: Data Structures/Stacks&Queues/StacksAndQueues.cpp

/**
========================================================================
Stacks & Queues

STACK: LIFO:    Implement a stack using a linked list.
                Using position 0 in a linked list we achieve:
                    add:    O(1)
                    remove: O(1)

QUEUE: FIFO:    Implement a queue using a linked list.
                    add:    O(1)
                    remove: O(1)
      
: zach wolpe
: zach.wolpe@medibio.com.au
: 02 July 2023
========================================================================
*/

#include <iostream>

class Node {
    public:
        int value;
        Node* next;

        Node(int value) {
            this->value = value;
            this->next  = nullptr;
        }
};

class Stack  {
    private:
        Node* top;
        int height;
    
    public:
        Stack(int value) {
            Node* node = new Node(value);
            top = node;
            int height = 1;
        }

        ~Stack() {
            Node* node = top;
            while (top) {
                top = top->next;
                delete node;
                node = top;
            }
        };

        void printData() {
            if (height) {
                Node* first = this->top;
                std::cout << "Stack : [";
                while (first) {
                    std::cout << "(" << first->value << ")";
                    if (first->next) std::cout << "->";
                    first = first->next;
                }
                std::cout << "]" << std::endl;
            }
        }

        void getTop() {
            if (top == nullptr) {
                std::cout << "Top: nullptr" << std::endl;
            }   else {
                std::cout << "Top: " << top->value << std::endl;
            }
        }

        void push(int value) {
            Node* node = new Node(value);
            node->next = top;
            top = node;
            height++;
        }

        int pop() {
            if (height == 0) return INT_MIN;
            Node* node = top;
            top = node->next;
            int pop_value = node->value;
            delete node;
            height--;
            return pop_value;
        }

        Node* pop_node() {
            if (height ==  0) return nullptr;
            Node* node = top;
            top = node->next;
            node->next = nullptr;
            return node;
        }


};



class Queue {
    private:
        Node* first;
        Node* last;
        int length;
    
    public:
        Queue(int value) {
            Node* node = new Node(value);
            this->first = node;
            this->last  = node;
            length = 1;
        }

        ~Queue() {
            // default destructor does not remove nodes.
            Node* temp = first;
            while (first) {
                first = first->next;
                delete temp;
                temp = first;
            }
        }

        void printData() {
            if (length) {
                Node* first = this->first;
                std::cout << "Stack : [";
                while (first) {
                    std::cout << "(" << first->value << ")";
                    if (first->next) std::cout << "->";
                    first = first->next;
                }
                std::cout << "]" << std::endl;
            }
        }

        void enqueue(int value) {
            Node* node = new Node(value);
            if (length == 0) {
                first = node;
                last  = node;
            } else {
                last->next = node;
                last = node; 
            }
            length++;
        }

        int dequeue() {
            if (length == 0) return INT_MIN;
            Node* node = first;
            int node_value = node->value;
            if (length == 1) {
                first = nullptr;
                last  = nullptr;
            } else first = node->next;
            length--;
            delete node;
            return node_value;
        }
};



int main() {
        
    Stack* stack = new Stack(19);
    stack->push(13);
    stack->printData();
    std::cout << "Pop: " << stack->pop();
    std::cout << "Pop: " << stack->pop();

    Queue* queue = new Queue(44);
    queue->enqueue(3);
    queue->enqueue(12);
    queue->enqueue(-64);


    queue->printData();
    std::cout << queue->dequeue() << std::endl;

            
}  
